src/Adapters/Redis/RedisStorage.php

```diff
@@ -0,0 +1,99 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Lamoda\Metric\Adapters\Redis;
+
+use Lamoda\Metric\Common\MetricSourceInterface;
+use Lamoda\Metric\Storage\MetricStorageInterface;
+use Lamoda\Metric\Storage\MutableMetricInterface;
+
+final class RedisStorage implements \IteratorAggregate, MetricStorageInterface, RedisStorageInterface
+{
+    /** @var RedisConnectionInterface */
+    private $redisConnection;
+
+    public function __construct(RedisConnectionInterface $redisConnection)
+    {
+        $this->redisConnection = $redisConnection;
+    }
+
+    public function getIterator(): \Traversable
+    {
+        return $this->getMetrics();
+    }
+
+    public function receive(MetricSourceInterface $source): void
+    {
+        $metrics = [];
+        foreach ($source->getMetrics() as $metric) {
+            $metrics[] = new MetricDto(
+                $metric->getName(),
+                $metric->resolve(),
+                $this->convertTagsForStorage($metric->getTags())
+            );
+        }
+        $this->redisConnection->setMetrics($metrics);
+    }
+
+    public function getMetrics(): \Traversable
+    {
+        $metricsData = $this->redisConnection->getAllMetrics();
+        foreach ($metricsData as $metricDto) {
+            yield $this->createMetricEntity(
+                $metricDto->name,
+                $metricDto->value,
+                $this->convertTagsFromStorage($metricDto->tags)
+            );
+        }
+    }
+
+    public function findMetric(string $name, array $tags = []): ?MutableMetricInterface
+    {
+        $value = $this->redisConnection->getMetricValue($name, $this->convertTagsForStorage($tags));
```

Comment:
This implementation always returns a value, so there is no way to say wheter we have this metric in the store or not

Reply:
It was a try to reduce a probabilty of a parallel setting of value without necessary of using locks.

For example, if two processes didn't found a metric at the same time, both of them will call setValue for the new metric. But if it is returned that metric virtually exists with value 0, in case of adjusting metric both processes will call adjustValue instead of setValue so both calls will not be lost.

Reply:
Take a look at this helper https://github.com/lamoda/php-metrics/blob/master/src/Storage/StoredMetricMutator.php
I think it already contains the logic you trying to achieve

Reply:
Well, it was exactly that class that led me to thoughts about possibility of losing values. Maybe I am wrong.

Added nullability.